### ISSUE TITLE
HTTP/2 support

### ISSUE THREAD

**Issue by olix0r** [BuoyantIO/linkerd#174]
This is a tracking ticket to enumerate all of the work needed to support HTTP/2 in (finagle and) linkerd.
- [x] cleartext http2 transport: https://github.com/twitter/finagle/commit/a067b4faa20713ce082376e6843334c100e7d3ad
- [x] support for http2settings negotiation
- [x] TLS support for netty4/http2: @vkostyukov ?
- [x] adapters between netty3 and netty4 http messages, necessary to wire http2 into finagle: @dschobel?
- [x] per @mosesn, https://github.com/netty/netty/pull/4503 is needed so that finagle can provide an http/2 server
- Many longer-term improvements are enumerated at https://github.com/netty/netty/issues/3667

**Comment by dschobel** [comment]
api is breaking on next monday's sync but the message mapping already landed https://github.com/twitter/finagle/blob/develop/finagle-http-netty4/src/main/scala/com/twitter/finagle/http-netty4/Bijections.scala

**Comment by olix0r** [comment]
It appears that netty/netty#4503 has made it into Netty-4.1.0.CR5. Finagle's next release

**Comment by olix0r** [comment]
There's experimental support for h2c (cleartext http/2) on the [ver/netty4 branch](https://github.com/BuoyantIO/linkerd/tree/ver/netty4).

```
:; /usr/local/bin/curl -svH 'Host: default' -o/dev/null --http2 localhost:4142       
* Rebuilt URL to: localhost:4142/
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 4142 (#0)
> GET / HTTP/1.1
> Host: default
> User-Agent: curl/7.48.0
> Accept: */*
> Connection: Upgrade, HTTP2-Settings
> Upgrade: h2c
> HTTP2-Settings: AAMAAABkAAQAAP__
> 
< HTTP/1.1 101 Switching Protocols
< connection: upgrade
< upgrade: h2c
< content-length: 0
* Received 101
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* TCP_NODELAY set
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
< HTTP/2.0 200
< content-type:text/html;charset=UTF-8
< content-length:2877
< l5d-dst-bound:/#/io.l5d.fs/default
< l5d-dst-path:/http/1.1/GET/default
< 
{ [2877 bytes data]
* Connection #0 to host localhost left intact
```

**Comment by olix0r** [comment]
@stevej and I have been slowly making progress on linkerd's HTTP/2 support.

Currently, we are observing the following issues with http2 _servers_:

HTTP/2 responses frequently fail:
```
*   Trying ::1...
* connect to ::1 port 4142 failed: Connection refused
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 4142 (#0)
> GET /admin/ping HTTP/1.1
> Host: localhost:4142
> User-Agent: curl/7.49.0
> Accept: */*
> Connection: Upgrade, HTTP2-Settings
> Upgrade: h2c
> HTTP2-Settings: AAMAAABkAAQAAP__
> 
< HTTP/1.1 101 Switching Protocols
< connection: upgrade
< upgrade: h2c
< content-length: 0
* Received 101
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* TCP_NODELAY set
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
* HTTP/2 stream 1 was not closed cleanly: CANCEL (err 8)
* Connection #0 to host localhost left intact
```

In linkerd, we see logs like:
```
I 0622 23:37:32.339 THREAD71 io.netty.handler.codec.http2.Http2FrameLogger.log: 
----------------OUTBOUND--------------------
[id: 0x4d96fd14, L:/127.0.0.1:4142 - R:/127.0.0.1:57569] RST_STREAM: streamId=1, errorCode=8
------------------------------------
E 0622 23:37:32.341 THREAD71: channelReadComplete: ChannelHandlerContext(finagleChannelTransport, [id: 0x02c2682d, L:/127.0.0.1:4142 - R:/127.0.0.1:57569])
W 0622 23:37:32.345 THREAD71 io.netty.channel.AbstractChannelHandlerContext.notifyOutboundHandlerException: Failed to fail the promise because it's done already: DefaultChannelPromise@59fab9dd(failure: io.netty.util.IllegalReferenceCountException: refCnt: 0)
io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
	at io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:111)
	at io.netty.buffer.AbstractDerivedByteBuf.release(AbstractDerivedByteBuf.java:65)
	at io.netty.handler.codec.http2.DefaultHttp2DataFrame.release(DefaultHttp2DataFrame.java:129)
	at io.netty.util.ReferenceCountUtil.release(ReferenceCountUtil.java:84)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.write(AbstractChannel.java:794)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.write(DefaultChannelPipeline.java:1271)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:724)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:716)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:802)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:709)
	at io.netty.channel.ChannelDuplexHandler.write(ChannelDuplexHandler.java:106)
	at com.twitter.finagle.netty4.channel.ChannelStatsHandler.write(ChannelStatsHandler.scala:79)
```

Furthermore "pure" (no prior knowledge) HTTP/2 requests are simply not processed by the server:
```
:; /usr/local/bin/curl --http2-prior-knowledge -vs localhost:4142/admin/ping
*   Trying ::1...
* connect to ::1 port 4142 failed: Connection refused
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 4142 (#0)
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* TCP_NODELAY set
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x7fae22004c00)
> GET /admin/ping HTTP/1.1
> Host: localhost:4142
> User-Agent: curl/7.49.0
> Accept: */*
> 
```
(There is no logging emitted in linkerd, no stats are incremented, and it appears that these messages do not make it out of netty, even).

**Comment by mosesn** [comment]
@olix0r is this bleeding finagle or release finagle?  I think the netty multiplexing codec has a framelogger on all the time, so you should be able to see something if you make it to there.

https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java#L119

**Comment by samuel** [comment]
This is great! I'm excited to see HTTP/2 is being worked on. Is it safe to assume that after HTTP/2 then gRPC specific support could follow? We (@sprucehealth) are eager to use linkerd but are waiting for gRPC.

**Comment by olix0r** [comment]
hey there, @samuel! that is indeed the plan. I think that once we get basic http2 support, grpc should basically Just Work through linkerd with minimal additional work.  We will probably end up building some additional grpc-specific utilities/configuration, as well.  But in the short term, we hope to nail down http2 support enough to allow grpc to be routed through linkerd.

**Comment by olix0r** [comment]
I've spent the past few days investigating finagle-http2. While it's known that finagle-http2 is not yet production ready, netty's codec-http2 library appears to have most of the necessary APIs to support HTTP/2

Initially, we want to implement an HTTP2 router in [linkerd](https://linkerd.io) that is capable of routing
[gRPC](http://grpc.io) messages. In order to provide the most basic level of gRPC transport, we need at least:

- Cleartext, prior-knowledge HTTP/2 server (that does not require an HTTP/1 message to initiate an HTTP/2 connection)
- Cleartext, prior-knowledge HTTP/2 client (that does not attempt to perform HTTP1-to-HTTP2 upgrading)

In order to satisfy these goals, we defer several requirements necessary for production-quality HTTP/2 support, including:

- HTTP1-to-HTTP2 client and server upgrading
- ALPN/TLS bootstrapping
- PUSH_PROMISE support -- we do not need to accept push requests from servers.
- Multiplexed clients -- it is sufficient to use a serial dispatcher (with multiple connections) as done with HTTP/1.
- Streaming -- while gRPC does require support for streaming, we can defer this for now.
- Trailers -- while gRPC does require support for trailers, we can defer this for now

### Changes ###

The [`ver/http2/mess` finagle branch](https://github.com/BuoyantIO/finagle/tree/ver/http2/mess) includes an altered finagle-http2 implementation.  The changes on this branch include:

- Http2ServerInitializer now _only_ installs Http2MultiplexCodec.
  This changes the server to only accept pure HTTP2 requests and not
  HTTP1 upgrade requests. Before, _only_ HTTP1 upgrade requests were
  supported.
- Http2Transporter now does not attempt to send HTTP1 upgrade
  requests. Before, _only_ HTTP1 upgrade requests were supported, but
  gRPC (at least, the go implementation) does not support HTTP1
  upgrade requests.
- Almost everything that happens on an http2 client or server is
  logged now. Multiple times, even. Lots and lots of logs.

linkerd now accepts grpc requests and handily routes them using the _io.l5d.path_ identifier:
```yaml
- protocol: http
  label: http2
  identifier:
    kind: io.l5d.path
  baseDtab: |
    /srv => /#/io.l5d.fs;
    /http => /srv;
  servers:
  - port: 4142
    ip: 0.0.0.0
    engine:
      kind: netty4
      http2: true
  client:
    engine:
      kind: netty4
      http2: true
```

However, the client side of finagle-http2 does not yet behave properly.  It appears that it does not properly initiate client connections:
- The connection preface isn't sent by the finagle-http2 client, causing the grpc server to error
with
```
2016/06/29 20:25:10 transport: http2Server.HandleStreams received bogus greeting from client: "\x00\x00\xb4\x01%\x00\x00\x00\x01\x00\x00\x00\x00\x0fD\x8bb\xbb\x0e\x93\xaf\xc8\xf6I"
```
It should be sending `PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n` (or, `0x505249202a20485454502f322e300d0a0d0a534d0d0a0d0a`)

in netty-4.1 there are two APIs for building HTTP2 clients, the Http2Connection API, which is currently used by finagle:
```scala
        val connection = new DefaultHttp2Connection(false /*server*/)

        val inboundH2ToH1 = new InboundHttp2ToHttpAdapterBuilder(connection)
          .maxContentLength(maxResponseSize)
          .propagateSettings(true)
          .build()

        new HttpToHttp2ConnectionHandlerBuilder()
          .connection(connection)
          .initialSettings(new Http2Settings().pushEnabled(false))
          .frameListener(inboundH2ToH1)
          .frameLogger(new Http2FrameLogger(LogLevel.INFO))
          .build()
```

As well as a newer, preferred but less mature `Http2MultiplexCodec`.  Unfortunately, there doesn't appear to be a good way to perform HttpObject<->Http2Frame object translation with this API, so it is difficult to just drop this into finagle-http2, since finagle-http2 still requires that a `Transport[HttpObject, HttpObject]` is exposed to the client stack.

### Next steps: ###

First and foremost, we need to get a proper cleartext, prior-knowledge HTTP/2 client configured properly.

Following that, we need to

- Streaming -- requires that DATA streams are written to a Message's Reader
- Multiplexing -- Requires replacing HttpClientDispatcher (a serial dispatcher) with a stream-aware dispatcher (as is done in Mux).
- Trailers -- requires extending Message types to include something like a `Future[Trailers]` that is satisfied when a message is complete.
- Servers should accept both Prior-Knowledge requests and HTTP1-upgrade requests.
- Clients should initiate either HTTP1-upgrade or Prior-Knowledge requests (as controlled by a Param)
- ALPN+TLS

It seems that it may be substantially easier to build out this functionality--especially multiplexing--by exposing an `Transport[Http2Frame, Http2Frame]` to an Http2-specific dispatcher.  It should be possible to map HTTP2 message types to com.twitter.finagle.http.Message types, but I'm not sure that this translation should happen on the netty channel.  It seems like this could be much easier to accomplish by translating types closer to the application code.

**Comment by mosesn** [comment]
Very cool!  I'd suggest you look at c.t.f.http.StreamTransport for the finagle.Message <=> netty.DataFrame adapter.  I don't think MultiplexCodec works on the client-side yet, fwiw: https://github.com/netty/netty/issues/4913.  It shouldn't be too bad to implement the Http2Frame converter when that happens though.

**Comment by olix0r** [comment]
I'm happy to report that we have a variant of finagle-http2 that is able to transmit gRPC messages!

I've modified an [example application](https://github.com/BuoyantIO/linkerd-examples/tree/ver/grpc) to use gRPC so that it works as follows:
- a _web_ service serves http/1 requests with chunked responses
- a _word_ service serves a simple gRPC service with a single endpoint.
- a _gen_ service serves a gRPC service with an endpoint that streams response data.
- _web_ is configured to connect its _word_ and _gen_ clients to a finagle server that routes requests to the appropriate backend.

```
:; curl -vs -o/dev/null 'localhost:8080/gob?limit=1000000'                
*   Trying ::1...
* Connected to localhost (::1) port 8080 (#0)
> GET /gob?limit=1000000 HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.49.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: text/plain
< Date: Thu, 14 Jul 2016 18:59:01 GMT
< Transfer-Encoding: chunked
< 
{ [3986 bytes data]
* Connection #0 to host localhost left intact
```
(... and it works, streaming large responses even! 🎉 )

While this work is based on finagle-http2, it diverged substantially. We have moved away from HTTP1 types entirely. No attempts have been made to preserve compatibility with finagle-http.  This approach required very little work on the server-side, as netty's `Http2Codec` handles most of the complexity of framing and demultiplexing -- we simply provide a `ServerStreamTransport` (analogous to finagle-http's `StreamTransport`) that maps H2 stream frames into our Request types and our Response types into H2 stream frames.

The client side of this is a bit more complex, since, as @mosesn mentioned, netty's `Http2MultiplexCodec` doesn't yet satisfy any of these concerns...  So, we introduced a [`ClientDispatcher`](https://github.com/BuoyantIO/linkerd/blob/ver/http2/router/http2/src/main/scala/com/twitter/finagle/buoyant/http2/ClientDispatcher.scala) to compensate for this.  The client dispatcher is responsible for initializing and managing outbound client streams and mapping our Request types into H2 stream frames (and mapping received H2 stream frames into an existing stream's response.  Furthermore, ClientDispatcher manages flow control by pushing window updates into netty's `Http2FrameCodec` (which handles buffering and delivery of window updates).

It was necessary to [patch netty](https://github.com/netty/netty/pull/5531) with a fix for window updates; but once I moved finagle onto a snapshot version of netty, everything worked as hoped (so this is all predicated on finagle using Netty 4.1.3).

This an exciting start. There are almost certainly severe bugs lurking in my integration, but it's nearly at a point where we can start iterating on specific parts of this. In the coming days, I'd like to prep the [`ver/http2` branch](https://github.com/BuoyantIO/linkerd/tree/ver/http2) so that I can start to share it more broadly for review.  Additionally, I'd like to get linkerd into a state where it's possible to configure an http2 router based on this codec.

We'll continue to update this issue as we progress.

**Comment by olix0r** [comment]
In the past two weeks, we've made substantial progress on our H2 implementation including:
- Simultaneous server support for H2C and HTTP1-to-HTTP2-upgrade
- Initial burn-in/perf baseline testing/profiling

In order for this code to hit linkerd's master branch, we're waiting for the next finagle release, which includes [Netty-4.1.3.Final](https://github.com/twitter/finagle/pull/522).  Furthermore, I'd like to support non-prior-knowledge (i.e. H1-to-H2-upgrade) client support so that linkerd can communicate with non-grpc, non-linkerd servers.  Following that, we'll be focused on ALPN/TLS and general feature parity (the H2 codec doesn't yet support tracing/context propagation/etc).

Summary: h2 works. we're close to merging it in linkerd. After that, we'll work with @mosesn and the finagle team to figure out the best path forward for graduating/merging our codec with that in finagle.

**Comment by mosesn** [comment]
I have a review up internally for clients that can upgrade or downgrade, but streaming still doesn't work (because of how the netty InboundHttp2ToHttpAdapter works).  I have some ideas for how to implement that, if you want to meet up at some point in time.  Good job on the server-side stuff–can you downgrade too?  Or does it only support http/2?

**Comment by olix0r** [comment]
@mosesn I haven't worked on H1 downgrading yet.  I've moved away from the finagle request types, but I believe that we should be able to be bridge the difference from [our (very abstract) request types](https://github.com/BuoyantIO/linkerd/blob/ver/h2%2Bupgrade/router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/Message.scala) to an implementation of c.t.f.h.Message.  We started on something that used Readers, but it turns out that Readers don't really model H2 streams as well as I'd like, so I've settled on a [DataStream](https://github.com/BuoyantIO/linkerd/blob/2533bc2de01263f8bb4afd8d09a279ee9b7bf605/router/h2/src/main/scala/com/twitter/finagle/buoyant/h2/Message.scala#L62) type that can be backed by a reader/writer, etc. (specifically, we want the ability to model a data frame with the EOS bit set -- something that Reader makes cumbersome).

I've tested streaming support in my branch at length and feel fairly happy with the implementation I've arrived on.

I'll be traveling next week, but when I get back it would be good to do a bit of a walkthrough and figure out how/if there's a path to merge codecs.

**Comment by olix0r** [comment]
Please excuse the radio silence.  This feature is still under active development/testing.

We've primarily been waiting for linkerd to be able to use a new finagle release.  We found a [show-stopper bug in 6.37](https://github.com/twitter/finagle/issues/549), which [we fixed](https://github.com/twitter/finagle/issues/550), but not before 6.38.0 was released.  We're now waiting on a 6.39.0 release, but I don't have a timeline for that.

@mosesn @kevinoliver @vkostyukov Can share any information 6.39.0 release timeline?

**Comment by mosesn** [comment]
@olix0r tentatively looks like first week of October.

**Comment by olix0r** [comment]
As we enter the first week of October, a brief update on the linkerd side of things:

* Prior-knowledge h2c works well enough to start testing against gRPC (without TLS).  I've [published a build](https://github.com/BuoyantIO/linkerd/releases/tag/h2-alpha-1) against Finagle HEAD that can be used for testing as we prepare this for a merge into master (following the next finagle release).

  Please check out this pre-release and give us early feedback. I expect (hope, even?) to find bugs here.

* I've started to work on [h2 (with TLS) support](https://github.com/BuoyantIO/linkerd/tree/ver/h2+alpn). Serverside TLS is working on that branch, with a few caveats. In working on this, I found a problem (h/t @moderation) with the way we configure netty4 server TLS.  I will fix this on master independently of the h2 work.

* I figured out how to squelch annoying log4j warnings. This will be fixed on master as well.

I think we're on track to merge nascent (explicitly experimental) h2 support as soon as finagle 6.39 is released (barring any quality issues in the release, but initial testing is promising).  There are a number of features that we'll want to follow up with after this release -- I'll enumerate these as we get closer to knowing what's going to make it in the next linkerd release.

**Comment by mosesn** [comment]
Cool!  On the finagle side, we have cleartext working, both h2c and prior knowledge (prior knowledge must be configured server-side–we're trying to work with netty https://github.com/netty/netty/issues/5857 to improve this, but folks don't seem terribly interested, so we might have to just write a hack into finagle).  We also have server-side http/2 with TLS over ALPN, but have not hooked up client-side http/2 with TLS.

We're planning on doing a release in the next two weeks.

**Comment by olix0r** [comment]
@mosesn for what it's worth, i tried to use your most recent finagle-http2 with h2c support, but gRPC expects that trailers are supported (and afaict, trailers are not supported by finagle), and so gRPC responses just never complete.  But I definitely borrowed heavily from your work, so thanks again for pushing forward with it.

In addition to trailers, I think we really need to support on end-to-end flow control, which finagle is not currently positioned to provide.  We should definitely continue to talk, though, about how we can unify implementations over time.

Let us know if there's anything we can do to help expedite a finagle release ;)  As I've noted before, we've been unable to get a stable upgrade since [early July](https://github.com/BuoyantIO/linkerd/commit/0c0ccfdc65b1bf70dd0962492d9b62e4ce16661e).

**Comment by mosesn** [comment]
@olix0r ah, if grpc needs trailers we'll need to add it to the finagle-http Message type.  Shouldn't be too hard other than that though.  👍  yeah, would love to standardize so we can get to a point where we can benefit from each other's work.

**Comment by buchgr** [comment]
@mosesn yes we require trailers to deliver status information for an RPC.

**Comment by olix0r** [comment]
The [ver/h2+h1up](https://github.com/BuoyantIO/linkerd/tree/ver/h2%2Bh1up) branch includes support for:
- h2 server (via _tls_ config)
- h2 client (via _tls_ config)
- h2c server (both prior knowledge and http/1 upgrade, simultaneously without config)
- h2c client with prior knowledge

Cleartext, http/1 upgrading clients are not yet supported, but this has no bearing on gRPC so this can land later.

Additionally, many of finagle-http's features---error handling, tracing, retries, per-request dtab overrides, etc---are not yet supported for linkerd's _h2_ protocol.  I'll continue work on feature parity.

In the meantime, I'll publish another pre-release with TLS support once I have a chance to write a README explaining the relevant configuration options.

**Comment by zackangelo** [comment]
Just to add a little detail around what @buchgr said, the most common way to convey error details for a failed RPC is in an HTTP/2 trailer. The gRPC java stack even does this by default if you craft the exception in the right way, see https://github.com/grpc/grpc-java/issues/681 for details.

**Comment by buchgr** [comment]
@zackangelo thanks for the additional details! I was actually refering to the gRPC protocol [1]. Trailers are _required_ by the gRPC protocol to transmit *gRPC* status codes. Even without your own *application level* status codes, one needs trailers for gRPC to work correctly.

The application level status codes you are refering to are the (optional) `*Custom-Metadata` in the response specification.

[1] https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses

**Comment by olix0r** [comment]
I've published another pre-release off of this newest work: https://github.com/BuoyantIO/linkerd/releases/tag/h2-alpha-2

Please don't hesitate to get in touch with questions/bugs/etc

**Comment by olix0r** [comment]
I've just published the [h2-alpha-3 pre-release](https://github.com/BuoyantIO/linkerd/releases/tag/h2-alpha-3).  This includes support for tls, linkerd headers (like l5d-dtab), proper proxy support (so that `http_proxy=... curl --http2 http://service/` works), and support for pluggable identifiers (the default, now, is to simply use the `:authority` header -- the _io.l5d.h2.path_ identifier is useful for routing grpc requests).

I think the big remaining tasks for linkerd's h2 implementation are to get parity for stats and tracing.  If we get a finagle release before that happens, we'll release h2 as-is and improve things in the next release.  (Though, it seems possible that this may be beat finagle 6.39 out the door).

**Comment by olix0r** [comment]
Whoops, sorry for dropping the thread on this issue.

linkerd's h2 support is fully released in linkerd-0.8.4, the protocol is still marked as _experimental_.  

There are a number of [enhancements](https://github.com/BuoyantIO/linkerd/labels/h2) planned for the future, but the basic features necessary for gRPC transport--both TLS and plaintext--work with no known bugs.

Furthermore, I've begun work on a gRPC frontend over our h2 implementation, so that it's easy to access/expose grpc services from linkerd and namerd.  The first use case for this will be to build a [gRPC interface for namerd](https://github.com/BuoyantIO/linkerd/issues/855).  Work on this is progressing well.

In order to [promote h2 to stable](https://github.com/BuoyantIO/linkerd/issues/854), we need your help battle-testing it.  If you run a linkerd with an h2 router, let us know!  If you _can't_ run linkerd with an h2 router due to some bug or lacking feature, _definitely let us know_.
